Implement support for most of the new string ABI

vedantk opened the pull request:
This teaches the string data formatter about some additional bits of the new ABI related to bridged NSStrings. Support for the 32-bit ABI is still WIP.

vedantk committed:
[CMake] Make check-lldb work with LLDB_CODESIGN_IDENTITY=''

On Darwin, if a test machine isn't set up for code-signing (see
docs/code-signing.txt), running check-lldb should use the system
debugserver instead of the unsigned one built in-tree. This makes it
possible to run lldb's test suite without having code-signing set up,
which is really convenient.

Differential Revision: https://reviews.llvm.org/D42215

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@322803 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry picked from commit 215f8dc9ada385e659d3b9d45d7df5b03e6f95db)

vedantk committed:
Remove commented-out debugging printf()'s. NFC.

vedantk committed:
Squelch compiler warnings in a test. NFC.

vedantk committed:
[DataFormatter] Delete dead swift string formatter code

vedantk committed:
[DataFormatter] Implement support for bridged NSStrings

This gets us most of the way towards understanding the new string ABI.
We're still missing support for 32-bit and ARM.

dcci commented on source/Plugins/Language/Swift/SwiftFormatters.cpp:
```diff
@@ -166,89 +166,6 @@ static bool invokeNSStringFormatter(Process &process, ValueObject &valobj,
-      StringPrinter::ReadStringAndDumpToStreamOptions());
-}
-
-bool lldb_private::formatters::swift::StringCore_SummaryProvider(
```
Whoah, thanks for removing all this cruft.

dcci commented on source/Plugins/Language/Swift/SwiftFormatters.cpp:
```diff
@@ -325,15 +242,7 @@ bool lldb_private::formatters::swift::StringGuts_SummaryProvider(
-    // stream.Printf("o (opaque) = %d\n", is_opaque);
-    // stream.Printf("w (width) = %d\n", is_utf_16);
-
-    if (!is_opaque && ((is_a_value && !is_Cocoa_or_small) ||
```
I didn't understand the whole check to begin with.
I trust you that you can simplify it, but, maybe, can we add a comment?

vedantk replied:
This checks for strings which aren't opaque or bridged. There is a "monster" comment above which documents why this class of strings need specific handling.

vedantk replied:
But it could use a comment like "Handle plain-old native Swift strings"

vedantk commented:
This fixes several test failures, but we can't land this at the moment because the bots are red.

dcci reviewed (commented):
Some comments inline. Not only this fixes several test failures, but also removes some code. Great work.

vedantk committed:
[DataFormatter] Add explanatory comments to the string formatter, NFC

vedantk commented:
Yes, the TestSwiftBridgedStringVariables.py I un-xfailed contains examples of each of these kinds of strings, and tests both the `fr v` and `fr v -O` code paths (i.e we test the path through the data formatter and the path through `stringForPrintObject`).

jimingham commented on packages/Python/lldbsuite/test/lang/swift/variables/bridged_string/main.swift:
```diff
@@ -13,12 +13,12 @@ import Foundation
-	var s5 = "abc" as NSString
-	var s6 = String(s5)
+	let s1 = "Hello world"
+	let s2 = "ΞΕΛΛΘ"
```
var's and let's are handled differently by the swift compiler.  I presume you changed this to remove the compiler warnings, but in tests like this we should really have a mix of var's and lets and we don't really care about the "you didn't use the var" warnings.  If the warnings bother you, just print all the variables at the end of the test...

Also, now that you know the String format, are each of the variants you had to cover below represented here?

vedantk replied:
Yes, I changed this to appease the compiler, but I'll try something like '-w' if it bothers me in the future.

All the variants we need to cover are taken care of, but we will need separate handling for the 32-bit ABI.

vedantk commented on packages/Python/lldbsuite/test/lang/swift/variables/bridged_string/main.swift:
```diff
@@ -13,12 +13,12 @@ import Foundation
-	var s6 = String(s5)
+	let s1 = "Hello world"
+	let s2 = "ΞΕΛΛΘ"
+	let s3 = s1 as NSString
```
@jimingham `s3` is the `is_Cocoa_or_small && !is_a_value` case

vedantk commented on packages/Python/lldbsuite/test/lang/swift/variables/bridged_string/main.swift:
```diff
@@ -13,12 +13,12 @@ import Foundation
+	let s3 = s1 as NSString
+	let s4 = s2 as NSString
+	let s5 = "abc" as NSString
+	let s6 = String(s5)
```
`s6` is the `is_Cocoa_or_small && is_a_value` case

jimingham replied:
It's a little weird that the test case has 6 variables in it but it only tests s3-s6.  s1 & s6 are probably the same, but we only test the NSString version of the non-ascii String (s2)  Does that one work?

vedantk replied:
Thanks for the catch! I didn't bother adding these checks because they are covered in another test, but it makes sense to have them for completeness. I've just pushed the change.

jimingham replied:
Oh, yes, while you are messing with this, lines 45-67 of that test case can be replaced with:
 
    	        (target, process, thread, bkpt) = lldbutil.run_to_source_breakpoint(self,
    	                                   "Set breakpoint here", self.main_source_spec)
 
Be nice to do this if possible whenever we touch a test case.  This is much easier to read and would have meant hundreds fewer test case changes in Adrian's work to move the build out of tree.

vedantk replied:
Done. And thanks, it's good to have this reminder!

jimingham reviewed (approved):
I didn't study the new layout of string, so I can't comment on how accurately this represents the new layout.  But provided the test feeds one of each of the cases you had to cover at the formatters, I don't doubt you got that right.

The code itself looks fine.

vedantk committed:
[DataFormatter] Extend a string formatting test case, NFC

This test case defines a few strings it never checks. Add checks for
them for completeness.

vedantk committed:
[DataFormatter] Simplify a test case, NFC

Use lldbutil.run_to_source_breakpoint to shorten a test case.

vedantk committed:
[test] Create the `swiftpr` test category and populate it

This test category is for Swift smoke tests run during pull-request
testing. It must only contain reliable, swift-related tests: failing,
expected-failing, unexpected-failing, or other flaky tests do not belong
in this category.

Initial results:

./swift/utils/build-script --release --skip-build-benchmarks
--skip-test-cmark --skip-test-swift --lldb --lldb-use-system-debugserver
--lldb-build-with-cmake --test --lldb-test-swift-only

Test Methods:        256
Reruns:                0
Success:              29
Expected Failure:      0
Failure:               0
Error:                 0
Exceptional Exit:      0
Unexpected Success:    0
Skip:                227
Timeout:               0
Expected Timeout:      0

vedantk commented:
Hi @shahmishal, this is the pull-request we discussed. It contains a data formatter fix for the new string ABI and a slew of test-infra fixes. Specifically, it creates the `swiftpr` test category we need for lldb pull-request testing. Could you force-merge this?

vedantk commented:
Thank you @shahmishal!
